File: backend/src/main/java/tom/projects/monkeylog/service/WorkoutService.java
package tom.projects.monkeylog.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import tom.projects.monkeylog.dto.workout.WorkoutCreateRequest;
import tom.projects.monkeylog.mapper.WorkoutMapper;
import tom.projects.monkeylog.model.workout.WorkoutType;
import tom.projects.monkeylog.model.workout.Workout;
import tom.projects.monkeylog.repository.workout.WorkoutRepository;
import tom.projects.monkeylog.security.AuthenticatedUser;

import java.time.LocalDateTime;
import java.util.Optional;

@Service
@Slf4j
@RequiredArgsConstructor
public class WorkoutService {
    private static final String WORKOUT_NOT_FOUND = "Workout not found.";
    private final WorkoutRepository workoutRepository;
    private final WorkoutMapper workoutMapper;
    private final ProgramService programService;

    public Page<Workout> getWorkouts(WorkoutType workoutType, LocalDateTime after, Pageable pageable) {
        if (after != null) {
            return workoutRepository.findAllByWorkoutTypeAndUserIdAndStartDateAfter(workoutType, AuthenticatedUser.getId(), after, pageable);
        }

        return workoutRepository.findAllByWorkoutTypeAndUserId(workoutType, AuthenticatedUser.getId(), pageable);
    }

    public Workout getWorkout(Long id) {
        return workoutRepository.findById(id)
                .filter(AuthenticatedUser::isResourceOwner)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, WORKOUT_NOT_FOUND));
    }

    public Optional<Workout> getActiveWorkout() {
        return workoutRepository.findAllByWorkoutTypeAndUserId(WorkoutType.ACTIVE, AuthenticatedUser.getId()).stream()
                .findFirst();
    }

    public Workout startWorkout() {
        Workout workout = new Workout();
        workout.setName("New empty workout");

        return startWorkout(workout);
    }

    public Workout startWorkout(Long id) {
        return startWorkout(Workout.clone(getWorkout(id)));
    }

    private Workout startWorkout(Workout workout) {
        workout.setStartDate(LocalDateTime.now());
        workout.setWorkoutType(WorkoutType.ACTIVE);
        workout.setUserId(AuthenticatedUser.getId());

        clearActive();

        return workoutRepository.save(workout);
    }

    private void clearActive() {
        workoutRepository
                .deleteAll(workoutRepository.findAllByWorkoutTypeAndUserId(WorkoutType.ACTIVE, AuthenticatedUser.getId()));
    }

    public Workout complete() {
        Workout workout = workoutRepository.findAllByWorkoutTypeAndUserId(WorkoutType.ACTIVE, AuthenticatedUser.getId())
                .stream().findFirst()
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, WORKOUT_NOT_FOUND));

        workout.getExerciseGroups().forEach(
                exerciseGroup -> exerciseGroup.getExerciseRows().removeIf(exerciseRow -> !exerciseRow.getIsLifted()));
        workout.getExerciseGroups().removeIf(exerciseGroup -> exerciseGroup.getExerciseRows().isEmpty());
        workout.setEndDate(LocalDateTime.now());
        workout.setWorkoutType(WorkoutType.COMPLETED);

        return workoutRepository.save(workout);
    }

    public Workout cloneToTemplate(Long id) {
        Workout workout = getWorkout(id);
        Workout newWorkout = Workout.clone(workout);
        newWorkout.setName(workout.getName() + " - copy");
        newWorkout.setWorkoutType(WorkoutType.TEMPLATE);
        newWorkout.setUserId(AuthenticatedUser.getId());

        return workoutRepository.save(newWorkout);
    }

    public Workout save(WorkoutCreateRequest workoutRequest) {
        Workout workout = workoutMapper.workoutRequestToWorkout(workoutRequest);
        workout.setWorkoutType(WorkoutType.TEMPLATE);
        workout.setUserId(AuthenticatedUser.getId());
        workout.setProgramWeek(workoutRequest.getProgramWeekId() == null ? null : programService.getProgramWeek(workoutRequest.getProgramWeekId()));

        return workoutRepository.save(workout);
    }

    public void delete(Long id) {
        workoutRepository.delete(getWorkout(id));
    }
}

File: backend/src/main/kotlin/tom/com/monkeylog/service/WorkoutService.kt
package tom.com.monkeylog.service

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException
import tom.com.monkeylog.common.dto.IdProjection
import tom.com.monkeylog.dto.workout.WorkoutCreateRequest
import tom.com.monkeylog.mapper.toEntity
import tom.com.monkeylog.model.workout.ExerciseRow
import tom.com.monkeylog.model.workout.Workout
import tom.com.monkeylog.model.workout.WorkoutType
import tom.com.monkeylog.repository.WorkoutRepository
import tom.com.monkeylog.security.AuthenticatedUser
import java.time.Instant
import java.util.*
import java.util.function.Predicate.not


@Service
class WorkoutService(
    private val workoutRepository: WorkoutRepository,
    private val programService: ProgramService,
) {
    fun getPagedWorkoutIds(workoutType: WorkoutType, pageable: Pageable): Page<IdProjection> {
        return workoutRepository.findAllByWorkoutTypeAndUserId(
            workoutType,
            AuthenticatedUser.id,
            pageable
        )
    }

    fun getWorkouts(ids: List<UUID>): List<Workout> {
        return workoutRepository.findAllByIdIn(ids)
    }

    fun getWorkout(id: UUID): Workout {
        return workoutRepository.findById(id)
            .filter(AuthenticatedUser::isResourceOwner)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, WORKOUT_NOT_FOUND) }
    }

    fun getActiveWorkouts(): List<Workout> {
        return workoutRepository.findAllByWorkoutTypeAndUserId(
            WorkoutType.ACTIVE,
            AuthenticatedUser.id
        );
    }

    fun startWorkout(): Workout {
        return Workout(name = "New empty workout").run(this::startWorkout)
    }

    fun startWorkout(id: UUID): Workout {
        return getWorkout(id).clone().run(this::startWorkout)
    }

    private fun startWorkout(workout: Workout): Workout {
        workoutRepository.deleteAll(getActiveWorkouts())

        return workout.apply {
            startDate = Instant.now()
            workoutType = WorkoutType.ACTIVE
            userId = AuthenticatedUser.id
        }.let(workoutRepository::save)
    }

    fun complete(): Workout {
        val activeWorkout: Workout = getActiveWorkouts().firstOrNull()
            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, WORKOUT_NOT_FOUND)

        return activeWorkout.apply {
            endDate = Instant.now()
            workoutType = WorkoutType.COMPLETED
            exerciseGroups.forEach { it.exerciseRows.removeIf(not(ExerciseRow::lifted)) }
            exerciseGroups.removeIf { it.exerciseRows.isEmpty() }
        }.let(workoutRepository::save)
    }

    fun cloneToTemplate(id: UUID) = getWorkout(id).clone().let(workoutRepository::save)

    fun save(workoutRequest: WorkoutCreateRequest): Workout {
        return workoutRequest.toEntity().apply {
            workoutType = WorkoutType.TEMPLATE
            userId = AuthenticatedUser.id
            programWeek = workoutRequest.programWeekId?.let(programService::getProgramWeek)
        }.let(workoutRepository::save)
    }

    fun delete(id: UUID) {
        workoutRepository.delete(getWorkout(id))
    }

    companion object {
        const val WORKOUT_NOT_FOUND = "Workout not found."
    }
}